### PR TITLE
[Dev] Fix an issue causing ExecuteTask to do much more work than intended

### PR DESCRIPTION
**Pull request by Tishj**
This PR makes `ExecutePushInternal` respect the `max_chunks` set in the call to `PipelineExecutor::Execute(idx_t max_chunks)`

When running DuckDB inside an environment where the control over received signals (like SIGINT) is not instant (like Python or Postgres), we rely on ExecuteTask to return after doing a small amount of processing work.

In the DuckDB CLI this problem was not apparent because SIGINT is handled and `Interrupt()` is called instantly, which is respected by the execution.
This problem only appeared because the `interrupt` flag could not be set as the application was first waiting for ExecuteTask to return.

### CODE REVIEW

**Comment by Tishj** [comment]
This is definitely out of my depth and I'm not confident all cases are considered or the necessary state in the PipelineExecutor is preserved to support pausing this process.

The original problem was that ExecutePushInternal would only return when the `Execute(chunk, chunk, idx_t)` call would not return HAVE_MORE_OUTPUT

**Comment by Tishj** [comment]
`TryFlushCachingOperators` is not prepared to deal with `ExecutePushInternal` returning `HAVE_MORE_OUTPUT`, that's the main thing I see that still needs to be fixed

**Comment by Tishj** [comment]
> `TryFlushCachingOperators` is not prepared to deal with `ExecutePushInternal` returning `HAVE_MORE_OUTPUT`, that's the main thing I see that still needs to be fixed

I fixed this by detecting that `HAVE_MORE_OUTPUT` gets returned from `ExecutePushInternal`, then returning `false` **without setting `remaining_sink_chunk`** to indicate that the flushing did not complete yet.

Outside of `TryFlushCachingOperator` we check if `remaining_sink_chunk` was set:
```c++
				if (remaining_sink_chunk) {
					return PipelineExecuteResult::INTERRUPTED;
				} else {
					return PipelineExecuteResult::NOT_FINISHED;
				}
```

By not setting `remaining_sink_chunk` we don't try to re-sink the chunk, enter TryFlushCachingOperators as normal, and because `in_process_operators` is non-empty (otherwise `HAVE_MORE_OUTPUT` would not be returned) we will get back into `ExecutePushInternal` where we continue pushing the produced chunks into the Sink.

File: src/include/duckdb/parallel/pipeline_executor.hpp [hunk 1]
```diff
@@ -32,6 +32,25 @@ enum class PipelineExecuteResult {
...
+	}
+
+public:
+	bool Next() {
+		if (processed >= maximum_to_process) {
+			return false;
+		}
+		processed++;
+		return true;
+	}
```
Comment (lnkuiper) [inline src/include/duckdb/parallel/pipeline_executor.hpp:47]: Can we add this method here?
```c++
bool IsDepleted() const {
    return processed >= maximum_to_process;
}
```
And use it in the `Next()` function (and possibly some assertions as well?).

Also, should `processed` ever be greater than `maximum_to_process`? Maybe we should assert that it is less or equal in `IsDepleted`.

File: src/parallel/pipeline_executor.cpp [hunk 2]
```diff
@@ -89,20 +90,33 @@ bool PipelineExecutor::TryFlushCachingOperators() {
...
-		if (push_result == OperatorResultType::BLOCKED) {
+		switch (push_result) {
+		case OperatorResultType::BLOCKED: {
 			remaining_sink_chunk = true;
 			return false;
-		} else if (push_result == OperatorResultType::FINISHED) {
+		}
+		case OperatorResultType::HAVE_MORE_OUTPUT: {
+			// The chunk budget was used up, pushing the chunk through the pipeline created more chunks
+			// we need to continue this the next time Execute is called.
```
Comment (lnkuiper) [inline src/parallel/pipeline_executor.cpp:108]: Can we assert that the chunk budget was depleted here using the method I proposed above? I'd like to be sure that this comment is correct

**Review by lnkuiper — CHANGES_REQUESTED** [review]
Thanks for the PR! This looks great. I just want to add a few assertions:

**Review by lnkuiper — APPROVED** [review]
Thanks for the changes! LGTM now

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**